softlayer/resource_softlayer_subnet.go

```diff
@@ -0,0 +1,386 @@
+					}
+					return
+				},
+			},
+
+			// IP version 4 or IP version 6
+			"ip_version": {
+				Type:     schema.TypeInt,
+				Optional: true,
+				Default:  4,
+				ForceNew: true,
+				ValidateFunc: func(v interface{}, k string) (ws []string, errs []error) {
+					ipVersion := v.(int)
+					if ipVersion != 4 && ipVersion != 6 {
+						errs = append(errs, errors.New(
+							"ip version should be either 4 or 6."))
+					}
+					return
+				},
+			},
+
+			"capacity": {
+				Type:     schema.TypeInt,
+				Required: true,
+				ForceNew: true,
+			},
+
+			// vlan_id should be configured when type is "Portable"
+			"vlan_id": {
+				Type:          schema.TypeInt,
+				Optional:      true,
+				Computed:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"endpoint_ip"},
+			},
+
+			// endpoint_ip should be configured when type is "Static"
+			"endpoint_ip": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ForceNew:      true,
+				ConflictsWith: []string{"vlan_id"},
+			},
+
+			// Provides IP address/netmask format (ex. 10.10.10.10/28)
+			"subnet_cidr": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"notes": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+		},
+	}
+}
+
+func resourceSoftLayerSubnetCreate(d *schema.ResourceData, meta interface{}) error {
+	sess := meta.(ProviderConfig).SoftLayerSession()
+
+	// Find price items with AdditionalServicesSubnetAddresses
+	productOrderContainer, err := buildSubnetProductOrderContainer(d, sess)
+	if err != nil {
+		return fmt.Errorf("Error creating subnet: %s", err)
+	}
+
+	log.Println("[INFO] Creating subnet")
+
+	receipt, err := services.GetProductOrderService(sess).
+		PlaceOrder(productOrderContainer, sl.Bool(false))
+	if err != nil {
+		return fmt.Errorf("Error during creation of subnet: %s", err)
+	}
+
+	Subnet, err := findSubnetByOrderId(sess, *receipt.OrderId)
+	if err != nil {
+		return fmt.Errorf("Error during creation of subnet: %s", err)
+	}
+
+	d.SetId(fmt.Sprintf("%d", *Subnet.Id))
+
+	return resourceSoftLayerSubnetUpdate(d, meta)
+}
+
+func resourceSoftLayerSubnetRead(d *schema.ResourceData, meta interface{}) error {
+	sess := meta.(ProviderConfig).SoftLayerSession()
+	service := services.GetNetworkSubnetService(sess)
+
+	subnetId, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return fmt.Errorf("Not a valid subnet ID, must be an integer: %s", err)
+	}
+
+	subnet, err := service.Id(subnetId).Mask(SubnetMask).GetObject()
+	if err != nil {
+		return fmt.Errorf("Error retrieving a subnet: %s", err)
+	}
+
+	if *subnet.AddressSpace == "PRIVATE" {
+		d.Set("private", true)
+	} else if *subnet.AddressSpace == "PUBLIC" {
+		d.Set("private", false)
+	}
+
+	d.Set("type", *subnet.SubnetType)
+	if strings.Contains(*subnet.SubnetType, "STATIC") {
+		d.Set("type", "Static")
+	} else if strings.Contains(*subnet.SubnetType, "VLAN") {
+		d.Set("type", "Portable")
+	}
```

Comment:
@renier Users may try to import different type of subnets. The current code will not generate an error. If you want to check the type strictly, you can throw an exception.